cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/CoreSchedulerService.java

```diff
@@ -116,6 +121,34 @@ protected void shutDown() throws Exception {
     }, co.cask.cdap.common.service.RetryStrategies.exponentialDelay(200, 5000, TimeUnit.MILLISECONDS));
   }
 
+  // Attempts to remove all jobs that are in PENDING_LAUNCH state.
+  // These are jobs that were about to be launched, but the scheduler shut down or crashed after the job was marked
+  // PENDING_LAUNCH, but before they were actually launched.
+  // This should only be called at startup.
+  private void cleanupJobs() {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          JobQueueDataset jobQueue = Schedulers.getJobQueue(context, datasetFramework);
+          try (CloseableIterator<Job> jobIter = jobQueue.fullScan()) {
+            LOG.info("Cleaning up jobs in state {}.", Job.State.PENDING_LAUNCH);
+            while (jobIter.hasNext()) {
+              Job job = jobIter.next();
+              if (job.getState() == Job.State.PENDING_LAUNCH) {
+                LOG.warn("Removing job {}.", job);
```

Comment:
Please add a more descriptive message. Perhaps: "Removing job {} from scheduler job queue because it was left from a previous run of the scheduler and its state is unknown."